### PR TITLE
Revise reordering to align with buckting

### PR DESCRIPTION
**Pull request by ruisizhang123**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #19
* #18
* __->__ #17
* #16

### CODE REVIEW

File: torch/_inductor/simple_fsdp/reorder.py [hunk 1]
```diff
@@ -68,53 +18,76 @@ def reorder_all_gather(
...
+                (
+                    (
+                        node_to_type[snodes[idx + 1]] == NodeType.AG_WAIT
+                        and node_to_type[snodes[idx + 2]] == NodeType.ALL_GATHER
+                    )
+                    or (
+                        node_to_type[snodes[idx + 1]] == NodeType.ALL_GATHER
+                        and node_to_type[snodes[idx - 1]] != NodeType.AG_WAIT
+                    )
+                )
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp/reorder.py:54]: I think you are trying to cover the cases with/without bucketing, but I can't parse this. Let's discuss in person before I make any suggestions.

File: torch/_inductor/simple_fsdp/reorder.py [hunk 2]
```diff
@@ -68,53 +18,76 @@ def reorder_all_gather(
...
+                (
+                    (
+                        node_to_type[snodes[idx - 1]] == NodeType.AG_WAIT
+                        and node_to_type[snodes[idx + 1]] == NodeType.ALL_GATHER
+                    )
+                    or (
+                        node_to_type[snodes[idx + 1]] == NodeType.ALL_GATHER
+                        and node_to_type[snodes[idx - 1]] != NodeType.AG_WAIT
+                    )
+                )
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp/reorder.py:74]: ditto

File: torch/_inductor/simple_fsdp/reorder.py [hunk 3]
```diff
@@ -123,20 +96,22 @@ def reorder_reduce_scatter(
     result_list: List[scheduler.BaseSchedulerNode] = []
     wait_list: List[scheduler.BaseSchedulerNode] = []
     node_to_type: Dict[scheduler.BaseSchedulerNode, int] = {}
-
     inverse_users, node_users = compute_node_users(snodes)
-
+    types = []
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp/reorder.py:100]: this is not used, except assigned on line 103

File: torch/_inductor/simple_fsdp/reorder.py [hunk 4]
```diff
@@ -123,20 +96,22 @@ def reorder_reduce_scatter(
...
+        if node_type in [NodeType.ALL_GATHER, NodeType.COMPUTE, NodeType.AG_WAIT]:
+            if node not in result_list and node not in wait_list:
                 result_list.append(node)
-        elif node_type == NodeType.WAIT:
-            if node_to_type[snodes[idx - 1]] == NodeType.REDUCE_SCATTER:
+        elif node_type == NodeType.RS_WAIT:
+            if node_to_type[snodes[idx - 1]] == NodeType.REDUCE_SCATTER or (
+                node_to_type[snodes[idx - 2]] == NodeType.REDUCE_SCATTER
+                and node_to_type[snodes[idx - 1]] == NodeType.RS_WAIT
+            ):
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp/reorder.py:114]: ditto

File: torch/_inductor/simple_fsdp/reorder.py [hunk 5]
```diff
@@ -153,25 +128,47 @@ def reorder_reduce_scatter(
 
     if len(wait_list) > 0:
         result_list.extend(wait_list)
+
+    # minor adjust to ensure overlapping between FWD & BWD
```
Comment (tianyu-l) [inline torch/_inductor/simple_fsdp/reorder.py:132]: not easy to understand, need to chat +1

**Review by tianyu-l — COMMENTED** [review]
In general looks ok, let's discuss what your algorithm means before further suggestions. E.g. maybe it'd be better to have separate node types for AG and AG copy-in. Let's see.

**Review by tianyu-l — APPROVED** [review]
As discussed offline, we should utilize the notion of `GroupedSchedulerNode` to simplify (or in some sense) revert the complicated changes in this PR.

We can land this PR for now, and do that as a follow up. Please address remaining minor comments in the PR before merging.

**Comment by tianyu-l** [comment]
> we should utilize the notion of GroupedSchedulerNode to simplify (or in some sense) revert the complicated changes in this PR.

Let's add a TODO item in the code

**Comment by ruisizhang123** [comment]
> As discussed offline, we should utilize the notion of `GroupedSchedulerNode` to simplify (or in some sense) revert the complicated changes in this PR.
> 
> We can land this PR for now, and do that as a follow up. Please address remaining minor comments in the PR before merging.

added

**Pull request merged**